### DISCUSSION TITLE
Automation ideas - new automations and their use cases

### DISCUSSION THREAD

**Discussion by mike12345567** [Budibase/budibase#628] [Ideas]
Wanted to open a discussion topic on new automations that could be of use to user's and specifically what their use case might be.

Some examples might be:

1. **Integration**: Twilio **Use case**: An app which sends a text message to the user when their order is ready.
2. **Integration**: Slack **Use case**: A chat bot that sends a message when some one changes the status of an opportunity in their CRM.

Would love people's input on this, throw out a few ideas so we can get an idea of the sort of automations we might need to build!

**Comment by mjashanks** [comment]
1. Google sheets is a big one- sheets has so many integrations with other systems.

If you can get data in and out of Google sheets, then it can go anywhere else. It's a good workaround for everything else.

2. Zapier / Integromat / IFTTT

Again - do to the size of their ecosystem, it opens up many other integrations out of the box

**Reply by mike12345567** [reply]
I think OAuth integration is going to be big for us, if we can build that in and support user's actually signing into their various external accounts and giving permission then I think we'll be able to do cool stuff with Google sheets!

**Comment by joebudi** [comment]
Two new automations; 1 simple, 1 advanced.

Slack / Teams / Email notifications are popular requests:
Record is created/updated/deleted > notification sent.

Advanced automation and wishful thinking:
DocuSign is another popular request:
Record updated (lead to customer) > contract created > contract uploaded to DocuSign > DocuSign emails new customer > customer signs contract > contract is updated in database > CRM figures updated > sales person notified.

**Reply by mike12345567** [reply]
The advanced one here is actually a pretty interesting example as its something that I don't think any platform can do currently, pause the workflow until a user completes an action.

Might be worth investigating how we could handle that in the future as I think that would be an incredibly powerful tool, especially if we have automation logging/history, can check on the state of the docusigning while the automation is still in progress/waiting.

**Reply by aptkingston** [reply]
That advanced example might be doable by splitting into 2 parts: 
1. Record updated **(trigger)** > contract created > contract uploaded to DocuSign > DocuSign emails new customer
2. Customer signs contract **([docusign webhook](https://www.docusign.com/blog/dsdev-adding-webhooks-application) - trigger)** > contract is updated in database > CRM figures updated > sales person notified.

We'll definitely be able to handle advanced automations like this once we have a webhook trigger implemented.

**Comment by joebudi** [comment]
Oh, I have thought of another and it ties in with one of my favourite tools - Typeform.

Public typeform updated > updates database

**Reply by mike12345567** [reply]
Integromat has a whole bunch of typeform stuff you can do, although thats obviously going from Budibase to Typeform: heres the list:
![image](https://user-images.githubusercontent.com/4407001/94026955-41502980-fdb2-11ea-89b1-f056f163f2c5.png)

As per issue #495 when we build up external web hooks to trigger workflows from other systems in Budibase we'll be able to use their [Webhook API](https://developer.typeform.com/webhooks/) to do all sorts of interesting things, once we have the external web hooks up and running we should try build your use case as everything will be there!

**Comment by aptkingston** [comment]
The essential ones that I think will open up a lot of possibilities:

**1. Webhook Trigger**
Being able to trigger a workflow by a webhook - allowing any external application to call in to Budibase, rather than us just calling out. This will not only allow complex data structures to be passed in, but also allow all sorts of callbacks from other applications.

**2. Find Record**
Search a table in the database for a specific record, based on knowing some fields. I think this is essential for workflows which will be creating records - even for stuff like checking if this customer already exists, or pulling customer details from the customer ID which was just created with a new order.
